ethereum_bridge/src/storage/proof.rs

```diff
@@ -0,0 +1,82 @@
+//! Proofs over some arbitrary data.
+
+use std::collections::BTreeMap;
+
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use namada_core::types::address::Address;
+use namada_core::types::key::{common, secp256k1};
+use namada_core::types::storage::BlockHeight;
+
+/// Ethereum proofs contain the [`secp256k1`] signatures of validators
+/// over some data to be signed.
+///
+/// At any given time, an [`EthereumProof`] will be considered
+/// "complete" once a number of signatures pertaining to validators
+/// reflecting more than 2/3 of the bonded stake on Namada is available.
+#[derive(Debug, BorshSerialize, BorshDeserialize, BorshSchema)]
+pub struct EthereumProof<T> {
+    /// The signatures contained in the proof.
+    pub signatures: BTreeMap<(Address, BlockHeight), secp256k1::Signature>,
+    /// The signed data.
+    pub data: T,
+}
+
+impl<T> EthereumProof<T> {
+    /// Return an incomplete [`EthereumProof`].
+    pub fn new(data: T) -> Self {
+        Self {
+            data,
+            signatures: BTreeMap::new(),
+        }
+    }
+
+    /// Add a new signature to this [`EthereumProof`].
+    pub fn attach_signature(
+        &mut self,
+        addr: Address,
+        height: BlockHeight,
+        signature: common::Signature,
+    ) {
+        if let common::Signature::Secp256k1(sig) = signature {
+            self.signatures.insert((addr, height), sig);
+        }
+    }
+
+    /// Add a new batch of signatures to this [`EthereumProof`].
+    pub fn attach_signature_batch<I>(&mut self, batch: I)
+    where
+        I: IntoIterator<Item = ((Address, BlockHeight), common::Signature)>,
```

Comment:
similarly here - think we should be using `secp256k1::Signature` instead of `common::Signature`